kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/credentials/Ec2MetadataCredentialsProvider.java

```diff
@@ -0,0 +1,318 @@
+            // current credential is expired, or it has been completed exceptionally.
+            // throw it away and generate a new one.
+            // we don't normally expect to follow the expired path as the preemptive refresh ought to have
+            // caused its refresh before its expiration.
+            current.compareAndSet(witness, null);
+            return getCredentials();
+        }
+
+        return witness.minimalCompletionStage();
+    }
+
+    private void scheduleCredentialRefresh(long delay) {
+        LOGGER.debug("Scheduling refresh of AWS credentials in {}ms", delay);
+
+        var refreshedCredFuture = new CompletableFuture<SecurityCredentials>();
+        executorService.schedule(() -> {
+            refreshCredential(refreshedCredFuture);
+            refreshedCredFuture.thenApply(sc -> {
+                var previous = current.getAndSet(refreshedCredFuture);
+                // the previous future have been already complete, but for safety, complete it anyway.
+                Optional.ofNullable(previous).ifPresent(f -> f.complete(sc));
+                return null;
+            });
+        }, delay, TimeUnit.MILLISECONDS);
+    }
+
+    private boolean isExpired(CompletableFuture<SecurityCredentials> witness) {
+        if (witness.isDone() && !witness.isCompletedExceptionally()) {
+            try {
+                return Optional.ofNullable(witness.getNow(null))
+                        .map(SecurityCredentials::expiration)
+                        .map(exp -> systemClock.instant().isAfter(exp))
+                        .orElse(false);
+            }
+            catch (CancellationException | CompletionException e) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private void refreshCredential(CompletableFuture<SecurityCredentials> future) {
+        getToken()
+                .thenCompose(tokenResponse -> client.sendAsync(createSecurityCredentialsRequest(tokenResponse.body()), HttpResponse.BodyHandlers.ofByteArray()))
+                .thenApply(Ec2MetadataCredentialsProvider::checkResponseStatus)
+                .thenApply(HttpResponse::body)
+                .thenApply(this::toSecurityCredentials)
+                .thenApply(this::checkSuccessfulState)
+                .whenComplete((credentials, t) -> propagateResultToFuture(credentials, t, future));
+    }
+
+    private void propagateResultToFuture(SecurityCredentials credentials, Throwable t, CompletableFuture<SecurityCredentials> target) {
+        final long nextRefresh;
+        if (t != null) {
+            LOGGER.warn("Refresh of EC2 credentials failed. Is IAM role {} assigned to this EC2 instance?", config.iamRole(), t);
+            tokenRefreshErrorCount.incrementAndGet();
+            target.completeExceptionally(t);
+
+            nextRefresh = systemClock.instant().plusMillis(backoff.backoff(tokenRefreshErrorCount.get())).toEpochMilli();
```

Comment:
This looks incorrect, the successful refresh path is doing `expiry - now` . In the unit test with logs up I see:

```
2024-12-16 11:03:57 DEBUG io.kroxylicious.kms.provider.aws.kms.credentials.Ec2MetadataCredentialsProvider:160 - Scheduling refresh of AWS credentials in 1734300238879ms
```

should this just be the backoff value?